# **Ejercicios de pair programming Módulo 3 Sprint 1**
## **Regresión lineal: Lección 7**
### **ANOVA**

---

En el pair programming de hoy usaremos el set de datos que guardastéis en el pair programming de normalización y estandarización.

Hasta ahora habéis estado evaluando las características de vuestro set de datos y habéis hecho una gran exploración, es el momento de hacer vuestro primer ANOVA! En el ejercicio de hoy tendréis que hacer un ANOVA con vuestro datos y hacer una interpretación de los resultados.

📌 NOTA Puede que vuestros datos no se ajusten o no cumplan todas las asunciones, no pasa nada, haced el ANOVA e interpretad los resultados. En próximas lecciones aprenderemos que podemos hacer cuando nos encontramos en esta situación.

In [1]:
# Tratamiento de datos
# -----------------------------------------------------------------------
import numpy as np
import pandas as pd
import random 

# Gráficos
# ------------------------------------------------------------------------------
import matplotlib.pyplot as plt
import seaborn as sns

# Estadísticos
# ------------------------------------------------------------------------------
import statsmodels.api as sm
from statsmodels.formula.api import ols
from statsmodels.multivariate.manova import MANOVA
from sklearn.preprocessing import StandardScaler

# Para navegar entre directorios
# -----------------------------
import os

plt.rcParams["figure.figsize"] = (10,8)

In [2]:
df_socio = pd.read_csv('../datos/esperanza_vida_transformado.csv', index_col=0)
df_socio.head(2)

country        region  population_in_millions  population_density  \
0  Argentina  SouthAmerica               -0.178879           -0.273196   
1  Australia       Oceania               -0.262734           -0.283298   

   urban_population_y  gdp_per_capita  health_total_expenditure  \
0            1.103154       -0.439011                 -1.075299   
1            0.771296        1.362030                  0.630170   

   health_care_index quantity_of_physicians  quality_of_life_index  \
0           0.742566                 Muchos               0.131610   
1           0.837385                 Medios               1.015861   

   purchasing_power_index  safety_index  pollution_index  \
0               -0.436682     -1.655219         0.057166   
1                1.027351     -0.217071        -1.285658   

   pop_using_improved_drinking_water  life_expectancy_at_birth_total_bc  
0                           0.474870                       1.596496e+11  
1                           0.710275                       2.611737e+11

In [3]:
anova = ols("""life_expectancy_at_birth_total_bc ~ region + population_in_millions + population_density +
       urban_population_y + gdp_per_capita + health_total_expenditure +
       health_care_index + quantity_of_physicians +
       quality_of_life_index + purchasing_power_index + safety_index +
       pop_using_improved_drinking_water + pollution_index""", data = df_socio).fit()

sm.stats.anova_lm(anova)

df        sum_sq       mean_sq  \
region                             13.0  1.425217e+23  1.096321e+22   
quantity_of_physicians              3.0  1.300712e+22  4.335706e+21   
population_in_millions              1.0  1.311924e+22  1.311924e+22   
population_density                  1.0  1.267122e+22  1.267122e+22   
urban_population_y                  1.0  2.449249e+21  2.449249e+21   
gdp_per_capita                      1.0  2.067059e+22  2.067059e+22   
health_total_expenditure            1.0  2.828289e+21  2.828289e+21   
health_care_index                   1.0  4.227577e+21  4.227577e+21   
quality_of_life_index               1.0  3.619119e+21  3.619119e+21   
purchasing_power_index              1.0  3.372267e+19  3.372267e+19   
safety_index                        1.0  7.028114e+20  7.028114e+20   
pop_using_improved_drinking_water   1.0  1.093002e+21  1.093002e+21   
pollution_index                     1.0  4.965631e+20  4.965631e+20   
Residual                           38.0  2.672282e+22  7.032321e+20   

                                           F        PR(>F)  
region                             15.589747  2.120339e-11  
quantity_of_physicians              6.165399  1.607476e-03  
population_in_millions             18.655635  1.083312e-04  
population_density                 18.018553  1.356809e-04  
urban_population_y                  3.482846  6.973991e-02  
gdp_per_capita                     29.393696  3.535089e-06  
health_total_expenditure            4.021844  5.207416e-02  
health_care_index                   6.011639  1.891955e-02  
quality_of_life_index               5.146408  2.905953e-02  
purchasing_power_index              0.047954  8.278348e-01  
safety_index                        0.999402  3.237790e-01  
pop_using_improved_drinking_water   1.554254  2.201387e-01  
pollution_index                     0.706116  4.059927e-01  
Residual                                 NaN           NaN

In [4]:
anova.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                                    OLS Regression Results                                   
=============================================================================================
Dep. Variable:     life_expectancy_at_birth_total_bc   R-squared:                       0.891
Model:                                           OLS   Adj. R-squared:                  0.813
Method:                                Least Squares   F-statistic:                     11.45
Date:                               Thu, 11 May 2023   Prob (F-statistic):           3.14e-11
Time:                                       18:00:45   Log-Likelihood:                -1659.5
No. Observations:                                 66   AIC:                             3375.
Df Residuals:                                     38   BIC:                             3436.
Df Model:                                         27                                         
Covariance Type:                           nonrobust                                         
=======================================================================================================
                                          coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------------------------
Intercept                            1.751e+11    3.1e+10      5.647      0.000    1.12e+11    2.38e+11
region[T.EasternAsia]                8.812e+10   3.51e+10      2.510      0.016     1.7e+10    1.59e+11
region[T.EasternEurope]              1.638e+09    3.2e+10      0.051      0.959   -6.31e+10    6.63e+10
region[T.NorthernAfrica]             1.205e+10    4.5e+10      0.268      0.790    -7.9e+10    1.03e+11
region[T.NorthernAmerica]           -3.348e+10   4.02e+10     -0.834      0.410   -1.15e+11    4.78e+10
region[T.NorthernEurope]            -5.677e+09   3.32e+10     -0.171      0.865   -7.28e+10    6.15e+10
region[T.Oceania]                    2.081e+09   3.92e+10      0.053      0.958   -7.72e+10    8.14e+10
region[T.South-easternAsia]          5.299e+10   3.61e+10      1.466      0.151   -2.02e+10    1.26e+11
region[T.SouthAmerica]               2.153e+10    3.1e+10      0.695      0.491   -4.12e+10    8.42e+10
region[T.SouthernAfrica]            -8.814e+10   4.67e+10     -1.889      0.067   -1.83e+11    6.31e+09
region[T.SouthernAsia]               3.745e+10   3.51e+10      1.066      0.293   -3.36e+10    1.09e+11
region[T.SouthernEurope]             3.541e+10   3.18e+10      1.115      0.272   -2.89e+10    9.97e+10
region[T.WesternAsia]                3.208e+10   3.14e+10      1.021      0.314   -3.15e+10    9.57e+10
region[T.WesternEurope]              2.478e+09    3.5e+10      0.071      0.944   -6.83e+10    7.33e+10
quantity_of_physicians[T.Muchos]    -2.987e+08   1.13e+10     -0.026      0.979   -2.32e+10    2.26e+10
quantity_of_physicians[T.Muy pocos] -1.513e+10   1.71e+10     -0.886      0.381   -4.97e+10    1.94e+10
quantity_of_physicians[T.Pocos]     -2.837e+09   1.25e+10     -0.227      0.822   -2.82e+10    2.25e+10
population_in_millions              -3.565e+09   5.59e+09     -0.638      0.527   -1.49e+10    7.75e+09
population_density                    8.87e+09   5.03e+09      1.764      0.086   -1.31e+09    1.91e+10
urban_population_y                   1.636e+09   6.56e+09      0.249      0.804   -1.16e+10    1.49e+10
gdp_per_capita                       2.846e+10   7.81e+09      3.645      0.001    1.27e+10    4.43e+10
health_total_expenditure             7.717e+09   6.85e+09      1.126      0.267   -6.16e+09    2.16e+10
health_care_index                    3.874e+09   5.63e+09      0.688      0.496   -7.53e+09    1.53e+10
quality_of_life_index                2.112e+10   1.24e+10      1.709      0.096   -3.91e+09    4.62e+10
purchasing_power_index              -4.299e+09   9.29e+09     -0.463      0.646   -2.31e+10    1.45e+10
safety

`Podemos observar que el R cuadrado se sitúa en un 0.891, lo que indica que el 89.3% de nuestra variable respuesta es causado por nuestras variables predictoras.
Todos los valores de nuestra columna PR(>F) están por debajo de 0.05, por lo que podemos asumir que todas las variables predictoras afectan a la variable respuesta.` 